Core: Support incremental compute for partition stats

ajantha-bhat opened the pull request.

ajantha-bhat commented:
Some engines want to synchronously write the partition stats. (Similar to how Trino synchronously write the puffin files during insert). Reading all the manifests in a table can be avoided to compute partition stats if the we compute the stats incrementally and merge it with previous stats.

ajantha-bhat commented:
Some engines want to synchronously write the partition stats. (Similar to how Trino synchronously write the puffin files during insert). Reading all the manifests in a table can be avoided to compute partition stats if the we compute the stats incrementally and merge it with previous stats.

@aokolnychyi, @pvary, @deniskuzZ, @gaborkaszab : Let me know what you guys think.

ajantha-bhat commented on data/src/main/java/org/apache/iceberg/data/PartitionStatsHandler.java:
```diff
@@ -150,6 +161,75 @@ public static PartitionStatisticsFile computeAndWriteStatsFile(Table table, long
+      PartitionMap<PartitionStats> statsMap = PartitionMap.create(table.specs());
+      // read previous stats
+      try (CloseableIterable<PartitionStats> oldStats =
+          readPartitionStatsFile(statsFileSchema, Files.localInput(statisticsFile.path()))) {
```
since the new unified tuple is used for reading the old stats file. It automatically handled the schema evolution.

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -53,14 +56,43 @@ public static Collection<PartitionStats> computeStats(Table table, Snapshot snap
+            .filter(
+                manifestFile ->
+                    snapshotIdsRange.contains(manifestFile.snapshotId())
+                        && !manifestFile.hasExistingFiles())
```
Note that not including the existing records manifest for incremental compute as those stats will be previously written when they are added.

deniskuzZ commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -53,14 +56,43 @@ public static Collection<PartitionStats> computeStats(Table table, Snapshot snap
-        .throwFailureWhenFinished()
-        .executeWith(ThreadPools.getWorkerPool())
-        .run(manifest -> statsByManifest.add(collectStats(table, manifest, partitionType)));
+    return collectStats(table, manifests, partitionType).values();
```
why not return list right away?

deniskuzZ commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -53,14 +56,43 @@ public static Collection<PartitionStats> computeStats(Table table, Snapshot snap
+   * @param currentSnapshot the snapshot till which partition stats is computed (inclusive).
+   * @return the {@link PartitionMap} of {@link PartitionStats}
+   */
+  public static PartitionMap<PartitionStats> computeStatsIncremental(
```
why do we need both? can't we always try to do incremental if prev snapshot stats file exists?

pvary replied:
I think it is good to have a way to force recompute stats, in case somehow the previous version of the stats become corrupted.

WDYT?

ajantha-bhat replied:
Yes. I kept the full computation to force recompute if the stats are corrupted.

deniskuzZ replied:
agree, ````ALTER TABLE t COMPUTE STATS```` could be exercised for that. 
However, why not add just an extra param like 'forceRecompute' instead of diff method signatures?

ajantha-bhat replied:
> However, why not add just an extra param like forceRecompute/recompute instead of diff method signatures?

Its a public interface added before 1.8.0. So, modifying it will cause API breakage. Hence, new interface. 
But they both reuse the same code once the manifest list is identified.

deniskuzZ replied:
we can overload a method and add an extra param, the previous one would use force flag

ajantha-bhat replied:
I am checking how to simplify now.

deniskuzZ replied:
I was thinking about something like this, and since we don't have dependency on `data` here, we could do this in `PartitionStatsHandler` 
````
public static Collection<PartitionStats> computeStats(Table table, Snapshot snapshot) {
  return computeStats(table, snapshot, true);
}
public static Collection<PartitionStats> computeStats(Table table, Snapshot snapshot, boolean recompute) {
 ....
  if (!recompute && latestStatsFile != null) {
    load incremental
  }
}
````

ajantha-bhat replied:
Not recommended to use boolean argument for public interface

https://iceberg.apache.org/contribute/?h=contribute#boolean-arguments

I have some refactoring in mind. I will finish in an hour and push it.

deniskuzZ replied:
few lines lower
````
  // GOOD: these uses of dropTable are clear to the reader
  dropTable(identifier, true /* purge data */);
  dropTable(identifier, purge);
````

deniskuzZ replied:
WDYM it's not public - that is the Catalog API

ajantha-bhat replied:
I was looking for a public keyword :) 

But yeah, better to follow the guidelines as mentioned above. It is more cleaner. 
I still refactored this to accept `boolean` in a private method of `PartitionStatsHandler`

deniskuzZ commented on data/src/main/java/org/apache/iceberg/data/PartitionStatsHandler.java:
```diff
@@ -149,6 +162,92 @@ public static PartitionStatisticsFile computeAndWriteStatsFile(Table table, long
+    Map<Long, PartitionStatisticsFile> stats =
+        partitionStatisticsFiles.stream()
+            .collect(Collectors.toMap(PartitionStatisticsFile::snapshotId, file -> file));
+    for (Snapshot snapshot : SnapshotUtil.ancestorsOf(snapshotId, table::snapshot)) {
```
could be simplified, no need to pass lookup fn

ajantha-bhat replied:
can you elaborate? 

I don't see any method for `ancestorsOf` that doesn't accept the lookup fn.

deniskuzZ replied:
sorry, my bad, I was looking at `oldestAncestorOf`

deniskuzZ commented on data/src/main/java/org/apache/iceberg/data/PartitionStatsHandler.java:
```diff
@@ -149,6 +162,92 @@ public static PartitionStatisticsFile computeAndWriteStatsFile(Table table, long
+
+    StructType partitionType = Partitioning.partitionType(table);
+    Schema statsFileSchema = schema(partitionType);
+    PartitionStatisticsFile statisticsFile = latestStatsFile(table, snapshotId);
```
I think that should be extracted to `collectStats`. `computeAndWriteStats` should call `collectStats` and write it to FS

ajantha-bhat replied:
`collectStats` is a `core` module API which doesn't have dependency on `data` module (`PartitionStatsHandler`). 

`PartitionStatsHandler` is responsible for reading the stats file (including old stats). So, this logic of finding the old stats and reading is kept here instead of core module API.

deniskuzZ replied:
Could we extract `collectStats` functionality within the `PartitionStatsHandler`. This method is already overloaded

ajantha-bhat replied:
refactored with whatever possible on data module.

deniskuzZ commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -45,22 +49,54 @@ private PartitionStatsUtil() {}
+   * @param currentSnapshot the snapshot till which partition stats is computed (inclusive).
+   * @return the {@link PartitionMap} of {@link PartitionStats}
+   */
+  public static PartitionMap<PartitionStats> computeStats(
```
`fromSnapshot`?

deniskuzZ commented on data/src/main/java/org/apache/iceberg/data/PartitionStatsHandler.java:
```diff
@@ -135,20 +147,110 @@ public static PartitionStatisticsFile computeAndWriteStatsFile(Table table) thro
+    return incrementalComputeAndWrite(table, snapshotId, false /* recompute */);
+  }
+
+  private static PartitionStatisticsFile incrementalComputeAndWrite(
```
not sure if we should have 'incremental' in the method name since it might not be incremental

deniskuzZ commented on data/src/main/java/org/apache/iceberg/data/PartitionStatsHandler.java:
```diff
@@ -135,20 +147,110 @@ public static PartitionStatisticsFile computeAndWriteStatsFile(Table table) thro
+    StructType partitionType = Partitioning.partitionType(table);
+    PartitionMap<PartitionStats> resultStatsMap;
+
+    if (recompute) {
```
Could we extract `resultStatsMap` construction into a separate method? This could simplify testing as well

ajantha-bhat commented:
I have added 1.9.0 milestone for this PR as it is a small change (excluding refactoring) and we still have some time for 1.9.0 sue to open issues in milestone.

pvary commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -45,22 +49,53 @@ private PartitionStatsUtil() {}
+      manifestFilePredicate =
+          manifestFile ->
+              snapshotIdsRange.contains(manifestFile.snapshotId())
+                  && !manifestFile.hasExistingFiles();
```
Don't we want this as a default predicate?
```
manifestFile -> !manifestFile.hasExistingFiles()
```

deniskuzZ replied:
we could add it as default filter:
````
List<ManifestFile> manifests =
  currentSnapshot.allManifests(table.io()).stream()
      .filter(manifestFilePredicate)
      .filter(manifestFile -> !manifestFile.hasExistingFiles())
      .collect(Collectors.toList());
````

ajantha-bhat replied:
good point. 

While computing incremental, I observed that it may become duplicate counts. So, I added. 
I do have some gaps, I need to understand fully when and all we mark manifest entry as existing.
Is there any scenario exist to consider "existing" entries or just "added" is enough? 
 
There is another check down below, that considers both added and existing (added long back). 
https://github.com/apache/iceberg/blob/d54d81ecc1968eb0ec16a1f266e1d0b29e53b2ef/core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java#L103

I will update the code to just keep added entry and also add a testcase of rewrite data files to ensure stats are same after the rewrite.

deniskuzZ replied:
what if we have compaction and expire snapshots? new manifests would have the EXISTING entries?

ajantha-bhat replied:
> I will update the code to just keep added entry and also add a testcase of rewrite data files to ensure stats are same after the rewrite.

I remember now why I don't have this testcase, compaction was only supported at spark layer and this testcase is in core/data module. So, I was waiting for spark procedure to add these test case.

ajantha-bhat replied:
ok. I spent some time. 

The base code behavior of keeping existing entries is also proper. Because when the user calls merge append. It marks old manifest as delete and creates a new manifest which has existing entries + new entries. So, we need to consider both.

deniskuzZ replied:
I don't completely understand why it should be any different, since for incremental processing, we only cover the snapshot range (e.g., [s2...s5]). 
If we had performed a recompute, we would have included all the manifest files.

ajantha-bhat replied:
I have removed the existing file level check from predicate and handled at entry level

```
if (entry.isLive()) {
          // Live can have both added and existing entries. Don't consider existing entries for
          // incremental compute as it was already included in previous compute.
          if (!incremental || entry.status() == ManifestEntry.Status.ADDED) {
            stats.liveEntry(file, snapshot);
          }
        } else {
          stats.deletedEntry(snapshot);
        }
```

This way existing test of both incremental and full compute will pass.

deniskuzZ replied:
compaction doesn't remove files. if we do snapshot expiry after compaction - stats is dropped as well so we start fresh.

deniskuzZ replied:
compaction doesn't remove the data. if we. expire S1 and S2 we don't have prev stats and start fresh (i.e. full compute)

pvary replied:
If we don't expire data, could we detect that S3 is only a compaction commit, and the stats don't need to be changed?

What if S3 instead is a MoW commit? Can we detect the changes and calculate stats incrementally?

ajantha-bhat replied:
We didn't need decrement stats for full compute because we were discarding the deleted manifests. Only considering live manifests. 

Now, I am not really sure for compaction, the current code will work. We may need decrement stats just for incremental compute. I will test compaction scenario tomorrow and handle this.

deniskuzZ reviewed (approved):
LGTM +1

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -45,22 +49,53 @@ private PartitionStatsUtil() {}
+   * @deprecated since 1.9.0, will be removed in 1.10.0; use {@link #computeStats(Table, Snapshot,
+   *     Snapshot)} instead.
    */
+  @Deprecated
```
I initially had first two method that you described.
But later wanted to optimize as @deniskuzZ suggested. 

I always end up little back and forth in user facing APIs 😄 

I don't see use case of exposing manifest level filter to user. Do you have something in mind?

deniskuzZ replied:
+1 on Peter's proposal. 
Since `PartitionStatsHandler` has a generic compute stats (handles both full and incremental compute) we only need single-purpose methods in PartitionStatsUtil and avoid passing NULLs. However, I am not sure if `Collection<PartitionStats>` return type is idea for the incremental case since we do an aggregation.

pvary replied:
@deniskuzZ is right about the return type - I have just copied 😄

ajantha-bhat replied:
I deprecated old method as the return type "collection" is useless for new logic. 

I added new methods as suggested.

deniskuzZ replied:
IDK, i would prefer to keep `Collection<PartitionStats> computeStats` instead of `PartitionMap<PartitionStats> computeStatsFull` - it just adds confusion.

ajantha-bhat replied:
If we check the `PartitionStatsHandler.computeAndWrite`, code is very clean if both the method returns `PartitionMap`? Plus we can always get the `Collection` by calling `values()`

So, easier to use the APIs  if both the interface has same return type.

deniskuzZ replied:
ok, but `PartitionStatsUtil` code is redundant and confusing now.

ajantha-bhat replied:
> ok, but PartitionStatsUtil code is redundant and confusing now.

Can you please elaborate what is confusing? (forget deprecated API), new APIs both return the PartitionMap<PartitionStats> after computing the stats. Which are easier to merge for incremental case with old stats in the caller. 

Returning collection and filling a map again to merge is not efficient memory wise and performance wise too.

deniskuzZ replied:
filling what map again?
````
Collection<PartitionStats> partitionStats;
    if (recompute) {
      partitionStats = PartitionStatsUtil.computeStatsFull(table, snapshot);
    } else {
      partitionStats = incrementalComputeAndMerge(table, snapshot, partitionType)values();
    }
    if (partitionStats.isEmpty()) {
      return null;
    }
    List<PartitionStats> sortedStats =
        PartitionStatsUtil.sortStats(partitionStats, 

Collection<PartitionStats> incrementalComputeAndMerge() {
  return map.values()
}
````

deniskuzZ replied:
```` 
// full recompute
 public static Collection<PartitionStats> computeStats(Table table, Snapshot snapshot) 

and

 // incremental recompute
 public static PartitionMap<PartitionStats> computeStatsIncremental(Table table, Snapshot fromSnapshot, Snapshot currentSnapshot) 
````

deniskuzZ replied:
Thanks @ajantha-bhat! Sorry for all this ping-pong.

ajantha-bhat replied:
check now. Maybe I didn't push :)

pvary commented on data/src/main/java/org/apache/iceberg/data/PartitionStatsHandler.java:
```diff
@@ -135,20 +146,112 @@ public static PartitionStatisticsFile computeAndWriteStatsFile(Table table) thro
+      return PartitionStatsUtil.computeStats(table, null, snapshot);
+    }
+
+    PartitionStatisticsFile statisticsFile = latestStatsFile(table, snapshot.snapshotId());
+    if (statisticsFile == null) {
+      LOG.info("Previous stats not found. Computing the stats for whole table.");
+      return PartitionStatsUtil.computeStats(table, null, snapshot);
+    }
```
Could this throw an error instead?

deniskuzZ replied:
why? That is to handle the case when no stats files existed before, and we need to execute a full computation.

pvary replied:
If I understand correctly, the user requested incremental stat compute, but with wrong parameters. In this case we could either "correct" the mistake or throw an error.

The question is, how frequent is the problem, and how easy is to detect from the user side

deniskuzZ replied:
what do you mean by wrong parameters? non existing snapshotId?

ajantha-bhat replied:
throwing error now and added the testcase

deniskuzZ replied:
i don't agree with that design, see https://github.com/apache/iceberg/pull/12629#discussion_r2016760966

pvary commented on data/src/main/java/org/apache/iceberg/data/PartitionStatsHandler.java:
```diff
@@ -135,20 +146,112 @@ public static PartitionStatisticsFile computeAndWriteStatsFile(Table table) thro
+    return record;
+  }
+
+  private static PartitionStatisticsFile latestStatsFile(Table table, long snapshotId) {
```
Do we have test for this code?
It is not too complicated, but non-trivial. Maybe at lease a test which depends on the logic would be good

ajantha-bhat replied:
added the testcase

ajantha-bhat commented:
@deniskuzZ, @pvary: Thanks guys for the review. I have addressed all the comment. You can take a fresh look again tomorrow :D (after some break :D)

ajantha-bhat commented on data/src/main/java/org/apache/iceberg/data/PartitionStatsHandler.java:
```diff
@@ -135,20 +142,114 @@ public static PartitionStatisticsFile computeAndWriteStatsFile(Table table) thro
+      Table table, Snapshot snapshot, StructType partitionType) throws IOException {
+    PartitionStatisticsFile statisticsFile = latestStatsFile(table, snapshot.snapshotId());
+    if (statisticsFile == null) {
+      throw new RuntimeException(
```
>  recompute flag loses it's purpose

there is no recompute flag exposed to the user. The private method (`incrementalComputeAndMerge`) which is throwing this exception is also always computing incrementally. 

> I would expect from computeAndWriteStatsFileIncremental do what's needed instead throwing Previous stats not found exception.

 `computeAndWriteStatsFileIncremental` says incremental compute. Forcefully recomputing when there is an error is not a good idea as the method's responsibility is just to try incremental compute? 
 
 Maybe I can expose another method called  `computeAndWriteStatsWithFallback()`, which will internally calls it?
 
 ```
public void computeAndWriteStatsWithFallback() {
    try {
        computeAndWriteStatsFileIncremental();
    } catch (RuntimeException e) {
        if ("bla-bla".equals(e.getMessage())) {
            computeAndWriteStats(); // Fallback in case of a specific error
        } else {
            throw e; // Re-throw unexpected errors
        }
    }
}
```

deniskuzZ replied:
I liked how you did it initially. 
Think about what changes are needed on the client side. I was planning just to replace the existing call to the incremental one unless it's ANALYZE TABLE. 
What are the use-cases when we would benefit from the stats file missing exception?

ajantha-bhat replied:
Lets see what @pvary thinks.

gaborkaszab replied:
I'm very late to this conversation, sorry about that :) I think we should talk a bit about how a user would use these APIs to compute stats and then we might be able to sort this disagreement out too.

I see [the other PR](https://github.com/apache/iceberg/pull/12451) to introduce a `compute_partition_stats` stored proc for the "full-compute" path. I'd assume there would be another proc, `incremental_compute_partition_stats` or similar that will execute the "incremental-compute" path. If my assumption is correct, I think the question is why would a user decide to call one instead of the other. The expectation here is that the "full-compute" path is more expensive than the "incremental-compute" path. So if the users motivation is to run this cheaper operation then falling back to the full compute could be misleading.
Or from a different angle: if the incremental path is expected to try first with the cheap computation and then fall back to the more expensive one, then what would be the point of having the `compute_partition_stats` procedure to execute "full-compute". Why would one call it? So in general I'm in favour of throwing an exception if the incremental computation is not feasible.

Well, if the plan is to have that single Spark procedure for both approaches, then everything I wrote above is irrelevant :)

deniskuzZ replied:
My take on this is to always use `incremental` unless you explicitly need to recompute. 
`incremental` should be smart enough to decide if it needs to start from scratch (first stats compute) or reuse prev stats, compute diff, and merge.

gaborkaszab replied:
I think there are different approaches that could work well. I understand that the most convenient would be to offer a single `computeStats` function that can decide between incremental or full computation. On the other hand that could hide some details from the users. I had the impression that on Iceberg APIs are designed in a way to have clear boundaries and not mix functionalities like incremental or full stat computation.
I believe that to come to a conclusion we might want to raise this question on dev@ to have wider visibility. People probably are busy with the upcoming Iceberg Summit, but still we could get good insights. @ajantha-bhat WDYT?

deniskuzZ replied:
@gaborkaszab, please check the code closely, `incremental` doesn't do `full computation` unless there were no prior stats. 
Maybe I am wrong, but creating a dev mail thread to discuss every minor change seems counterproductive to me, when we already have people interested in this change here.
Maybe it's worth dropping `computeAndWrite` since `recompute` flag introduces confusion.

ajantha-bhat replied:
> Maybe I am wrong, but creating a dev mail thread to discuss every minor change seems counterproductive to me, when we already have people interested in this change here.

Strongly agree.

nastra commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -40,27 +44,48 @@ public class PartitionStatsUtil {
-    Preconditions.checkArgument(table != null, "table cannot be null");
-    Preconditions.checkArgument(Partitioning.isPartitioned(table), "table must be partitioned");
-    Preconditions.checkArgument(snapshot != null, "snapshot cannot be null");
+    Preconditions.checkArgument(table != null, "Table cannot be null");
```
FYI, in other places we typically use `Invalid table: null`. The same for snapshot: `Invalid snapshot: null`

ajantha-bhat replied:
I see only in `TableUitl` the message you mentioned, in Flink, I saw some places it is similar to what is here. 

Since it is not standardized and we already shipped with that error message. I don't think we need to correct it.

ajantha-bhat replied:
updated to `Invalid snapshot: null `and `Invalid table: null` syntax

ajantha-bhat commented:
I am traveling for Iceberg summit and may not be able to work on it this week. 
I will be moving this to 1.10.0 milestone. This is a good optimization. But need some testing for compaction scenarios.

gaborkaszab commented on data/src/main/java/org/apache/iceberg/data/PartitionStatsHandler.java:
```diff
@@ -135,20 +143,113 @@ public static PartitionStatisticsFile computeAndWriteStatsFile(Table table) thro
+    PartitionMap<PartitionStats> statsMap = PartitionMap.create(table.specs());
+    // read previous stats, note that partition field will be read as GenericRecord
+    try (CloseableIterable<PartitionStats> oldStats =
+        readPartitionStatsFile(schema(partitionType), Files.localInput(statisticsFile.path()))) {
```
Just brainstorming: If the referenced stat file doesn't exist we throw some data reading specific exception here (haven't found a test to see what exception, but probably I wasn't searching hard enough). What I have in mind is that we could wrap it into an exception that is relevant for the incremental stats computing. Like `IncrementalComputeStatsFailedException` or such. Then we could include a member in this specific exception that holds the snapshot ID of `statisticsFile`. On the caller side then the caller could decide to fix this broken state of the table/stats and invoke compute stats for that particular snapshot.
LMK WDYT!

gaborkaszab commented on data/src/main/java/org/apache/iceberg/data/PartitionStatsHandler.java:
```diff
@@ -135,20 +143,113 @@ public static PartitionStatisticsFile computeAndWriteStatsFile(Table table) thro
+      }
+    }
+
+    throw new RuntimeException(
```
I'd introduce a new exception instead of this general one. See one of my comments above.

gaborkaszab commented on data/src/main/java/org/apache/iceberg/data/PartitionStatsHandler.java:
```diff
@@ -135,20 +143,113 @@ public static PartitionStatisticsFile computeAndWriteStatsFile(Table table) thro
+  static PartitionStatisticsFile latestStatsFile(Table table, long snapshotId) {
+    List<PartitionStatisticsFile> partitionStatisticsFiles = table.partitionStatisticsFiles();
+    if (partitionStatisticsFiles.isEmpty()) {
+      return null;
```
Not sure if this was discussed, but I think returning `null` here is not consistent with throwing an exception at L249. I'd prefer throwing an exception here too. On the callsite we anyway throw an exception.

gaborkaszab commented on data/src/main/java/org/apache/iceberg/data/PartitionStatsHandler.java:
```diff
@@ -135,20 +143,113 @@ public static PartitionStatisticsFile computeAndWriteStatsFile(Table table) thro
+    Map<Long, PartitionStatisticsFile> stats =
+        partitionStatisticsFiles.stream()
+            .collect(Collectors.toMap(PartitionStatisticsFile::snapshotId, file -> file));
+    for (Snapshot snapshot : SnapshotUtil.ancestorsOf(snapshotId, table::snapshot)) {
```
Thunking out loud here: What if we have a long list of snapshots and only the very first one has stats? Then we practically do an almost full recompute of stats in case we execute this for the latest one. Isn't it misleading in terms of expected performance? I mean if the users decised to run the incremental computation, they'd expect it to be cheaper than the full computation but in this case it isn't. I know, this is an edge case, not sure if we should do anything about it.
Maybe having a threshold config for the number of ancestors we'd want to visit? Just asking if it makes sense at all.

gaborkaszab reviewed (commented):
Great work, @ajantha-bhat ! Sorry, I'm a bit late. LMK if my comments make sense!

pvary commented:
Let's wait for @ajantha-bhat to come back from the Summit and see what he thinks.

Based on the discussion above we could just provide these 2 methods on the API:
```
/**
  * Updates the partition statistics for the table.
  * <ul>
  *    <li>If there are existing stats for the table then finds the latest one,
  *              and does incremental stats calculation from there.
  *    <li>If there are no current stats, calculate them from scratch
  * </ul>
  */
public static Collection<PartitionStats> computeStats(Table table) {
```
and
```
/**
  * Forcefully updates the partition statistics for the table. Calculates them from scratch 
  * and ignores previous stats. Could be used when stats corruption is suspected, or the 
  * incremental stats calculation is costly (too many snapshots since last calculation).
  */
public static Collection<PartitionStats> reComputeStats(Table table) {
```

pvary commented:
> Let's wait for @ajantha-bhat to come back from the Summit and see what he thinks.
> 
> Based on the discussion above we could just provide these 2 methods on the API:
> 
> ```
> public static Collection<PartitionStats> computeStats(Table table) {
> ```
> 
> and
> 
> ```
> public static Collection<PartitionStats> reComputeStats(Table table) {
> ```

Based on our offline discussion with @gaborkaszab, the incremental stats calculation doesn't need to traverse multiple files, so we only need to force recompute stats in cease of some stat corruption. I think that could be a different design/PR (either drop the corrupt stats, or force recompute).

Then here we just need an api like:
```
/**
  * Updates the partition statistics for the table.
  * <ul>
  *    <li>If there are existing stats for the table then finds the latest one,
  *              and does incremental stats calculation from there.
  *    <li>If there are no current stats, calculate them from scratch
  * </ul>
  */
public static Collection<PartitionStats> computeStats(Table table) {
```

gaborkaszab commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -40,27 +44,48 @@ public class PartitionStatsUtil {
+   * @param currentSnapshot the snapshot till which partition stats is computed (inclusive).
+   * @return the {@link PartitionMap} of {@link PartitionStats}
+   */
+  public static PartitionMap<PartitionStats> computeStatsIncremental(
```
I think this naming could be misleading: users might expect that calling this function will produce the stats ready to be used, but after checking the code I have the impression that this produces the stats diffs between 2 particular snapshots and won't create 'ready-to-use' stats. Users have to additionally merge the stats returned by this function with the stats from `fromSnapshot`.
As option a) can't we do that 'final' merge also within this function and not leave it to the caller?
option b) would be to clearly express that users get stat diff that they have to merge with the last known stats.

pvary commented:
> @pvary, i don't follow where you are proposing these API changes?
> [..]
> If you are proposing to implement these API in `PartitionStatsUtil`, it would be problematic since it doesn't have ref to data module, see [#12629 (comment)](https://github.com/apache/iceberg/pull/12629#discussion_r2013945601)

I try to wrap my head around the API.
Is there some specific reason to separate the 2 util classes? Do we expect someone just using the calculation and someone using the stat file generation too? Or it was just a technical limitation?
(Left a comment on the discussion linked above)

deniskuzZ commented:
> Do we expect someone just using the calculation and someone using the stat file generation too? Or it was just a technical limitation?
I think the original design tried to separate the functionality of stats evaluation and stats file management (stats read/write).  However, I wouldn't expect someone to use `PartitionStatsUtil` directly.

pvary commented:
> > Do we expect someone just using the calculation and someone using the stat file generation too? Or it was just a technical limitation?
> 
> @pvary, I think the original design tried to separate the functionality of stats evaluation and stats file management (stats read/write). However, I wouldn't expect someone to use `PartitionStatsUtil` directly. see #11216

After discussing this with @deniskuzZ, I feel that we probably would want to move the `PartitionStatsUtil` and the `PartitionStatsHandler` into the same package, and deprecate the methods from the `PartitionStatsUtil` (make them package private in the long run), to decrease the size of the API. What do you think @ajantha-bhat, @gaborkaszab?

ajantha-bhat commented:
> After discussing this with @deniskuzZ, I feel that we probably would want to move the PartitionStatsUtil and the PartitionStatsHandler into the same package, and deprecate the methods from the PartitionStatsUtil

If we can have all the functionality in the core, we could have had everything in the same package. The separation was because parquet reader and writer was not available in the core. Now that we have `InternalData`, let me re-evaluate about moving all the functionality to the core module by using `InternalData`.

ajantha-bhat commented:
I tried to move `PartitionStatsHandler` to `core` and use `InternalData`, I can't make test case run as it needs `iceberg-parquet` dependencies. I tried adding this to core test dependency. It will be a circular dependency. So, not allowed. 

Other way is to move `PartitionStatsUtil` to `data` module, but since computing stats is core functionality, I don't want to move it to data. 

So, we need to drop the idea of this movement. We need to work with having `PartitionStatsUtil` in core and `PartitionStatsHandler` in data, what can be done to keep interface simple. I will work it and also finish the compaction test.

pvary commented:
> I tried to move `PartitionStatsHandler` to `core` and use `InternalData`, I can't make test case run as it needs `iceberg-parquet` dependencies.

The tests could run with Avro in core.
How InternalData tested with Parquet? We could use the same approach with this as well

> Other way is to move `PartitionStatsUtil` to `data` module, but since computing stats is core functionality, I don't want to move it to data.

I have some concerns about the size of the `core` module. I have very bad experience in ever growing big modules. In Hive everything was in `ql` module, and the dependencies messed up. That is why, I was debating wether we could find a better place for this feature, but I don't yet have a strong feeling around this.

> So, we need to drop the idea of this movement. We need to work with having `PartitionStatsUtil` in core and `PartitionStatsHandler` in data, what can be done to keep interface simple. I will work it and also finish the compaction test.

I really want to avoid this. This will create public interfaces which are public just because we can not organize our code better.

ajantha-bhat commented:
> The tests could run with Avro in core.
How InternalData tested with Parquet? We could use the same approach with this as well

But we need to test parquet aswell. And I didn't see any test in core with InternalData and Parquet. Since parquet is expected to be added at the runtime dependency it will work in production. I am not sure, how are we gonna add test in core module for it as it is a circular dependency. 
cc: @rdblue

ajantha-bhat commented:
@pvary, @gaborkaszab, @deniskuzZ : I went ahead and made a bold changes as perter suggested in https://github.com/apache/iceberg/pull/12946. It will avoid too many public interface confusion. Once that PR is merged. I will rebase this PR.

ajantha-bhat commented:
@pvary, @gaborkaszab, @deniskuzZ: Today I spent some time about how the incremental stats should be used by end users. 

**By default, it should be incremental compute (but the incremental compute should do full compute if the table has no partition stats available previously). Other failure cases Incremental stats can throw error. 
User will also have an option to force the full compute.** 

`CALL catalog_name.system. compute_partition_stats('db.sample'); ` -- does incremental compute if previous stats exist, else full compute. 

`CALL catalog_name.system. compute_partition_stats(table => 'db.sample', full_compute => true); ` -- full compute (doesn't care about previous stats)

Do we have any difference of opinion of this? I didn't jump into code to avoid back and forth. The thing we should keep in mind that users should not have to change their script for first time full compute and incremental computes after that. That is why "the incremental compute should do full compute if the table has no partition stats available previously".

Also, I am waiting for parent PR (Internal data #12946 ) to be merged to rebase this PR.

ajantha-bhat commented:
> Do we need drop stats? If we have drop stats, then we might not need this?

@pvary: Partition stats is per snapshots and tracked by table metadata, when we force compute, if the stats present for pervious snapshots, it will be cleaned during `expire_snapshots`. If the stats exist for current snapshot and still we want to recompute, old stats will be made orphan for current snapshot and new stats will be registered for current snapshot. Orphan stats will be cleaned by `remove_orphan_files`.  

> We should consider other engines, like Trino, Hive, Impala too

It is a same API, so other engines can also follow this method. 

I will rebase and update the PR today.

gaborkaszab commented:
> CALL catalog_name.system.compute_partition_stats('db.sample');  -- does incremental compute if previous stats exist, else full compute.

I like this approach.

About the `full_compute => true` param, could you help what would be the user motivation of calling this version of the procedure? There is already one version that could decide between incremental or full compute, so this one seems unnecessary. Unless the use-case is that the user learns that some previous stats are broken and hence want to do a full recompute. But then if they know that stats are broken, I agree with Peter that they should drop stats and then use the other variant of the `compute_partition_stats` procedure without the `full_compute` param. Is there anything I miss?

Following this logic, we might not want to expose `PartitionStatsHandler.computeAndWriteStatsFileIncremental()` as public since the other `computeAndWriteStats` version should be able to figure out to do incremental or full compute.
Any thoughts?

ajantha-bhat commented:
@pvary, @gaborkaszab, @deniskuzZ, @nastra: PR is ready for review.

pvary commented:
Discussed this with @ajantha-bhat offline:
- We think that the need to recompute the stats is very-very rare. Ajantha mentioned that maybe when the stats file was deleted accidentally then it could lead to corruption when the stats removal is needed.
- Also he mentioned that using the java API the user could remove the stats already:
```
    table.refresh();
    UpdatePartitionStatistics update = table.updatePartitionStatistics();
    table.snapshots().forEach(s -> update.removePartitionStatistics(s.snapshotId()));
    update.commit();
```
- Based on this, if there are no immediate needs from the compute engines we could just omit the `computeAndWriteStatsFullRefresh` method.
- We can always add the new method to the API if the need rises.

What do you think @deniskuzZ, @gaborkaszab?

gaborkaszab commented:
> What do you think

Not introducing a specific API for the "force full computation" path makes sense to me.

Thanks for the heads-up on the "drop stats" topic. I think it's fine. We might want to introduce a parameterless `UpdatePartitionStatistics.removePartitionStatistics()` later on, but we'll see.

Thanks for your work @ajantha-bhat !

deniskuzZ commented:
Hive provides the QL to force the stats recompute. Maybe use-cases could be limited and mainly related to recovery, but still, I think it won't harm to have an API for that. 
We can work around it with `removePartitionStatistics` and then compute, but that would create 2 snapshots, not sure if that is a good approach.

pvary commented:
Do we know someone from Trino / Spark who can chime in with their requirements?

ajantha-bhat commented:
I have updated the PR with just one interface (as a new commit)

deniskuzZ commented:
@ajantha-bhat, sorry, I meant 2 metadata files, still those might not be small. 
Could you please explain what is the concern in adding the API to do the full re-compute, it would bring some confusing or polute the API? Why do we force clients to hack it? 
In Hive we need this feature, so what is the proposal - hack on a client or bring the code to the engine repo?

ajantha-bhat commented:
> Could you clarify the concern around keeping the API to trigger full partition stats recompute? Clients have to rely on workarounds, even though Iceberg internally supports this through a private method.

There are no strong concern. We felt it is redundant to have many APIs. Plus the reason for full compute again is very rare (maybe only during corruption).  Plus there are ways to achieve full compute with the single API by clearing stats.  

> PS: what's the iceberg view on the fact that we are changing the behavior of the existing API (full recompute -> incremental)? From a client's perspective, it might be considered a breaking change.

Still the full stats available for the user. The way it compute internally has changed. No difference in the output for the user.

> I just checked the Impala docs (https://impala.apache.org/docs/build/html/topics/impala_compute_stats.html) and they mentions support for both options:

This is little different. Per partition or whole table. (Not based on the snapshot)

@deniskuzZ: I have a question for hive users, if the user calls incremental first time (table without previous stats) are you expecting it to throw error or do full compute?

deniskuzZ commented:
> This is little different. Per partition or whole table. (Not based on the snapshot)
If the partition spec is not provided, stats is computed for all the partitions individually.

> if the user calls incremental first time (table without previous stats) are you expecting it to throw error or do full compute?
do a full compute.

pvary commented:
@ebyhr, @findepi: Do I remember correctly that you work on Trino?
We are debating of the api required by the engines for partition stats calculation.
We will have an API, which calculates the stats for a partition. It will either calculate from scratch, or if previous snapshots had some outdated stats, it will reuse stats for the old snapshot, and calculates the current stats incrementally.
Do you think Trino would need/use a different API which fully recalculates the stats?

Thanks,
Peter

ebyhr commented:
I believe @raunaqmorarka is the best person to answer the question.

pvary commented:
@deniskuzZ: How does Hive INCREMENTAL stats work? I'm not able to find the doc 😢 

About the Impala, here is what I have found:

> For a particular table, use either COMPUTE STATS or COMPUTE INCREMENTAL STATS, but never combine the two or alternate between them. If you switch from COMPUTE STATS to COMPUTE INCREMENTAL STATS during the lifetime of a table, or vice versa, drop all statistics by running DROP STATS before making the switch.

For me, this means that Impala uses incremental stats in a very different way than we do it in Iceberg. Having a full recompute would not help them, but they would need a "drop stats"

ajantha-bhat commented:
@raunaqmorarka: Do you have any opinion on whether Trino needs force refresh API for stats? 

@pvary and @deniskuzZ: I also checked Spark and they doesn't have incremental or force refresh option. One API should be enough. https://spark.apache.org/docs/latest/sql-ref-syntax-aux-analyze-table.html

I think there is no difference of opinion on the current code (one API).  We are just debating on whether another API of force refresh is needed. So, I think we can go ahead with this PR and later add force refresh option if really needed. This PR is hanging from long time. Merging this will enable the further work on spark action and other engine integrations.

deniskuzZ commented:
Please don't take my comment as a blocker for merging the PR. However, why not be a bit more flexible and retain the existing force refresh API for recovery, especially since it's already in use by some of the engines? 
We're not introducing a brand-new API; it's something that was already in place.

deniskuzZ commented on core/src/main/java/org/apache/iceberg/PartitionStatsHandler.java:
```diff
@@ -130,15 +151,28 @@ public static PartitionStatisticsFile computeAndWriteStatsFile(Table table) thro
+          "Using full compute as previous statistics file is not present for incremental compute.");
+      stats = computeStats(table, snapshot, file -> true, false /* incremental */).values();
+    } else {
+      stats = incrementalComputeAndMerge(table, snapshot, partitionType, statisticsFile);
```
maybe `incrementalStatsComputeAndMerge`

ajantha-bhat replied:
updated as `computeAndMergeStatsIncremental`

pvary commented on core/src/main/java/org/apache/iceberg/PartitionStats.java:
```diff
@@ -135,6 +135,38 @@ public void deletedEntry(Snapshot snapshot) {
+   *
+   * @param snapshot the snapshot corresponding to the deleted manifest entry.
+   */
+  public void deletedEntryForIncrementalCompute(ContentFile<?> file, Snapshot snapshot) {
```
The public methods should be deprecated first

ajantha-bhat replied:
done

pvary reviewed (approved):
Last minor changes

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/PartitionStatsHandler.java:
```diff
@@ -275,9 +386,17 @@ private static PartitionMap<PartitionStats> collectStats(
-          stats.liveEntry(file, snapshot);
+          // Live can have both added and existing entries. Don't consider existing entries for
+          // incremental compute as it was already included in previous compute.
+          if (!incremental || entry.status() == ManifestEntry.Status.ADDED) {
```
updated.

pvary replied:
thx

ajantha-bhat commented:
> Please don't take my comment as a blocker for merging the PR. However, why not be a bit more flexible and retain the existing force refresh API for recovery, especially since it's already in use by some of the engines?
We're not introducing a brand-new API; it's something that was already in place.

Looks like we (mainly Peter) have decided to go with just one API now. If there are strong requirements, we will add force refresh in the future.

ajantha-bhat commented:
PR is ready to be merged, I will rebase the spark actions and procedure PR once this is merged.

pvary commented:
Merged to main.
Thanks @ajantha-bhat for the PR, and @deniskuzZ, @nastra, @gaborkaszab for the reviews!

Here we went for the narrowest possible API. Feel free to revisit the need for the "invalidate"/"recompute" API from time-to-time, and if there are multiple engines requesting the feature we could add it.
